parquet-column/src/main/java/parquet/schema/Types.java

```diff
@@ -555,6 +628,349 @@ protected GroupType build(String name) {
+        parent.setKeyType(build("key"));
+        if (parent.parent != null) {
+          return new MapValueBuilder<P>(parent,
+              type).repetition(repetition);
+        } else {
+          return new MapValueBuilder<P>(parent,
+              parent.returnType, type).repetition(repetition);
+        }
+      }
+
+      public MapValueBuilder<P> requiredValue(PrimitiveTypeName type) {
+        return value(type, Type.Repetition.REQUIRED);
+      }
+
+      public MapValueBuilder<P> optionalValue(PrimitiveTypeName type) {
+        return value(type, Type.Repetition.OPTIONAL);
+      }
+
+      public MapGroupValueBuilder<P> value(Type.Repetition repetition) {
+        ((MapBuilder)parent).setKeyType(build("key"));
+        return new MapGroupValueBuilder<P>(parent).repetition(repetition);
+      }
+
+      public MapGroupValueBuilder<P> requiredGroupValue() {
+        return value(Type.Repetition.REQUIRED);
+      }
+
+      public MapGroupValueBuilder<P> optionalGroupValue() {
+        return value(Type.Repetition.OPTIONAL);
+      }
+    }
+
+    public static class MapGroupValueBuilder<P> extends BaseGroupBuilder<P,
+        MapGroupValueBuilder<P>> {
+      private final MapBuilder<P> parent;
+
+      public MapGroupValueBuilder(MapBuilder<P> parent) {
+        super(parent.parent);
+        this.parent = parent;
+      }
+
+      public P named(String name) {
+        parent.setValueType(build("value"));
+        return parent.named(name);
+      }
+
+      @Override
+      protected MapGroupValueBuilder<P> self() {
+        return this;
+      }
+    }
+
+    protected void setKeyType(Type keyType) {
+      this.keyType = keyType;
+    }
+
+    protected void setValueType(Type valueType) {
+      this.valueType = valueType;
+    }
+
+    private Type keyType = null;
+    private Type valueType = null;
+    private Class<P> returnType = null;
+
+    public MapBuilder(P parent) {
+      super(parent);
+    }
+
+    private MapBuilder(Class<P> returnType) {
+      super(returnType);
+      this.returnType = returnType;
+    }
+
+    @Override
+    protected MapBuilder<P> self() {
+      return this;
+    }
+
+    public MapKeyBuilder<P> key(PrimitiveTypeName type) {
+      if (parent != null) {
+        return new MapKeyBuilder<P>(this, type);
+      } else {
+        return new MapKeyBuilder<P>(this, returnType, type);
+      }
+    }
+
+    public MapGroupKeyBuilder<P> groupKey() {
+      return new MapGroupKeyBuilder<P>(this);
+    }
+
+    @Override
+    protected Type build(String name) {
+      if (keyType == null) {
+        keyType = STRING_KEY;
+      }
+      return buildGroup(repetition)
+          .as(OriginalType.MAP)
+          .repeatedGroup()
+              .addFields(keyType, valueType)
+              .named("map")
+          .named(name);
+    }
+
+    public P named(String name) {
+      Preconditions.checkState(originalType == null,
+          "Cannot call \"as\" for a MAP type.");
```

Comment:
I think a slightly better error message would be that MAP is already the logical type and you can't change it.